### ISSUE TITLE
dedicated status worker should be last to go into shutdown

### ISSUE THREAD

**Issue by voelzmo** [cloudfoundry/bosh#1917]
**Motivation**
BOSH drain scripts leave workers running until the current task is finished. In the case of a single long-running task blocking your shutdown, this means all workers currently not running a task are shut down while the worker currently executing a long-running task remains up.
This leads to the unfortunate situation where you cannot execute any commands that require a task: ssh `bosh ssh` into the VMs involved in the long-running tasks, `bosh logs` to look at their logs `bosh task --debug` to look at the debug log of the currently running task. 

**Proposal**
If a dedicated status worker exists, it should be the last one to shutdown. Only when all other workers are shutdown, we should shutdown the dedicated status worker. Also: a potential `read only` mode, where no tasks are picked up from the queue needs to respect status tasks such as `ssh` and getting a tasks debug logs.

**Comment by cppforlife** [comment]
@voelzmo im still not not convinced that this is somethign that could be done well in drain scripts in this particular case given that other components have to keep on running for this to be useful. what if nats job introduces drain script, etc.

i think a better solution is to have a CLI command that would rely on https://github.com/cloudfoundry/bosh/pull/1863 to stop picking up new tasks. once all tasks are done then this command could continue with delete-env. one update is done, command would unblock and director would start picking up enqueued tasks.

**Comment by mfine30** [comment]
This makes sense to me. Closing/re-opening to trigger a story

**Comment by mfine30** [comment]
We ended up actually doing the work for this in a separate [story](https://www.pivotaltracker.com/story/show/163869051). The solution we took is to effectively shut down workers in reverse-numerical order and consider worker 0 a prioirty worker. I gave it a go and it works as expected. 

One interesting thing I did notice is that `bosh logs` is not a priority task but `bosh ssh` is. I'm hoping to figure out what's what either offline or in a follow up

**Comment by voelzmo** [comment]
Hey @mfine30, I'm surprised this got prioritized suddenly and you already explored the solution space. What about @cppforlife's thoughts on the risk of implicitly depending on other components with this approach? How did you end up dealing with this?

**Comment by jfmyers9** [comment]
I think that the concerns brought up are still theoretically valid and might force us to reconsider this approach if we end up introducing logic that brings dependent services down in drain.

That being said, we chose to just progress with this approach as its a small quality of life improvement for most operators and doesn't have any negative impacts. I believe we also fixed some issues with the current drain implementation as well.
